File: common/src/main/scala/com/goexp/common/util/web/url/UrlBuilder.java
package com.goexp.common.util.web.url;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

public final class UrlBuilder {

    private final String host;

    private final Map<String, String> params = new HashMap<>();


    private UrlBuilder(final String host) {
        this.host = host;
    }

    public static UrlBuilder create(final String host) {
        return new UrlBuilder(host);
    }

    public final UrlBuilder param(final String name, final String value) {
        params.put(name, value);

        return this;
    }

    public final String build() {

        var queryString = params
                .entrySet()
                .stream()
                .map(entry -> String.format("%s=%s", entry.getKey(),
                        URLEncoder.encode(entry.getValue(), StandardCharsets.UTF_8)
                ))
                .collect(Collectors.joining("&"));


        return String.format("%s?%s", host, queryString);
    }
}

File: common/src/main/scala/com/goexp/common/util/web/url/UrlBuilder.scala
package com.goexp.common.util.web.url

import java.nio.charset.StandardCharsets.UTF_8

import scala.collection.mutable

object UrlBuilder {
  def apply(host: String): UrlBuilder = new UrlBuilder(host)
}

final class UrlBuilder private(val host: String) {
  private val params = mutable.Map[String, String]()

  def param(name: String, value: String): UrlBuilder = {
    params += name -> value
    this
  }

  def param(name: String, value: Int): UrlBuilder = {
    param(name, value.toString)
  }

  def build: String = {
    val queryString = params.to(LazyList)
      .map { case (k, v) => s"$k=${v.urlEncode(UTF_8)}" }
      .mkString("&")

    s"$host?$queryString"
  }
}